File: sample/src/main/java/com/glee/exoplayersample/ApeExtractor.java
package com.glee.exoplayersample;

import android.support.annotation.Nullable;

import com.google.android.exoplayer2.extractor.Extractor;
import com.google.android.exoplayer2.extractor.ExtractorInput;
import com.google.android.exoplayer2.extractor.ExtractorOutput;
import com.google.android.exoplayer2.extractor.ExtractorsFactory;
import com.google.android.exoplayer2.extractor.Id3Peeker;
import com.google.android.exoplayer2.extractor.PositionHolder;
import com.google.android.exoplayer2.metadata.Metadata;
import com.google.android.exoplayer2.metadata.id3.Id3Decoder;

import java.io.IOException;
import java.util.Arrays;

class ApeExtractor implements Extractor {
    public static final ExtractorsFactory FACTORY = new ExtractorsFactory() {

        @Override
        public Extractor[] createExtractors() {
            return new Extractor[] {new ApeExtractor()};
        }

    };
    /**
     * Returns whether this extractor can extract samples from the {@link ExtractorInput}, which must
     * provide data from the start of the stream.
     * <p>
     * If {@code true} is returned, the {@code input}'s reading position may have been modified.
     * Otherwise, only its peek position may have been modified.
     *
     * @param input The {@link ExtractorInput} from which data should be peeked/read.
     * @return Whether this extractor can read the provided input.
     * @throws IOException          If an error occurred reading from the input.
     * @throws InterruptedException If the thread was interrupted.
     */

    private static final byte[] FLAC_SIGNATURE = {'f', 'L', 'a', 'C', 0, 0, 0, 0x22};
    private final boolean isId3MetadataDisabled = true;
    private final Id3Peeker id3Peeker;
    private Metadata id3Metadata;

    /**
     * Peeks ID3 tag data (if present) at the beginning of the input.
     *
     * @return The first ID3 tag decoded into a {@link Metadata} object. May be null if ID3 tag is not
     *     present in the input.
     */
    @Nullable
    private Metadata peekId3Data(ExtractorInput input) throws IOException, InterruptedException {
        input.resetPeekPosition();
        Id3Decoder.FramePredicate id3FramePredicate =
                isId3MetadataDisabled ? Id3Decoder.NO_FRAMES_PREDICATE : null;
        return id3Peeker.peekId3Data(input, id3FramePredicate);
    }
    private boolean peekFlacSignature(ExtractorInput input) throws IOException, InterruptedException {
        byte[] header = new byte[FLAC_SIGNATURE.length];
        input.peekFully(header, 0, FLAC_SIGNATURE.length);
        return Arrays.equals(header, FLAC_SIGNATURE);
    }
    @Override
    public boolean sniff(ExtractorInput input) throws IOException, InterruptedException {
        if (input.getPosition() == 0) {
            id3Metadata = peekId3Data(input);
        }
        boolean b = peekFlacSignature(input);
        return b;
    }

    public ApeExtractor() {
        id3Peeker = new Id3Peeker();
    }

    @Override
    public void init(ExtractorOutput output) {
    }

    @Override
    public int read(ExtractorInput input, PositionHolder seekPosition) throws IOException, InterruptedException {
        return 0;
    }

    @Override
    public void seek(long position, long timeUs) {

    }

    @Override
    public void release() {

    }
}

File: sample/src/main/java/com/glee/exoplayersample/ApeExtractor.kt
package com.glee.exoplayersample

import com.glee.exoplayersample.ape.APEFileInfo
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.Format
import com.google.android.exoplayer2.ext.ffmpeg.FfmpegDecoder
import com.google.android.exoplayer2.ext.ffmpeg.FfmpegDecoderException
import com.google.android.exoplayer2.extractor.*
import com.google.android.exoplayer2.util.MimeTypes
import com.google.android.exoplayer2.util.Util
import java.io.IOException
import java.nio.charset.Charset


//class ApeHeader constructor(
//        var compressionLevel: Int = 0
//) {
//
//
//}

internal class ApeExtractor : Extractor {
    companion object {
        /**
         * The number of input and output buffers.
         */
        private const val NUM_BUFFERS = 16
        /**
         * The initial input buffer size. Input buffers are reallocated dynamically if this value is
         * insufficient.
         */
        private const val INITIAL_INPUT_BUFFER_SIZE = 960 * 6

        private const val TAG = "MAC"
    }

    private lateinit var output: ExtractorOutput
    private lateinit var track: TrackOutput
    private lateinit var decoder: FfmpegDecoder
    var apeFileInfo: APEFileInfo? = null

    @Throws(IOException::class, InterruptedException::class)
    override fun sniff(input: ExtractorInput): Boolean {
        val array = ByteArray(4)
        input.peekFully(array, 0, 4, true)
        val isApe = String(array, Charset.forName("US-ASCII")).trim() == "MAC"
        if (isApe) {
            apeFileInfo = APEFileInfo.read(input)
        }
        return isApe
    }

    override fun init(output: ExtractorOutput) {
//        val mediaFormat = Format.createAudioSampleFormat(null,
//                MimeTypes.AUDIO_RAW, null,
//                streamInfo.bitRate(),
//                streamInfo.maxDecodedFrameSize(),
//                streamInfo.channels,
//                streamInfo.sampleRate,
//                getPcmEncoding(streamInfo.bitsPerSample),
//                /* encoderDelay= */ 0,
//                /* encoderPadding= */ 0, null, null,
//                /* selectionFlags= */ 0, null,
//                if (isId3MetadataDisabled) null else id3Metadata)/* id= *//* codecs= *//* initializationData= *//* drmInitData= *//* language= */

        this.output = output
        track = output.track(0, C.TRACK_TYPE_AUDIO)
        apeFileInfo?.let {
            track.format(Format.createAudioSampleFormat(
                    null,
                    MimeTypes.AUDIO_RAW,
                    null,
                    it.nDecompressedBitrate,
                    it.nTotalFrames,
                    it.nChannels,
                    it.nSampleRate,
                    Util.getPcmEncoding(it.nBitsPerSample),
                    null, null, 0, null
            ))
        }
        output.endTracks()
        try {
            decoder = FfmpegDecoder(
                    NUM_BUFFERS,
                    NUM_BUFFERS,
                    INITIAL_INPUT_BUFFER_SIZE,
                    MimeTypes.AUDIO_APE,
                    null,
                    true
            )
        } catch (e: FfmpegDecoderException) {
            e.printStackTrace()
            throw e
        }
    }


    @Throws(IOException::class, InterruptedException::class)
    override fun read(input: ExtractorInput, seekPosition: PositionHolder): Int {

//        val byteArray = ByteArray(4)
//        input.read(byteArray, 0, 4)
//        Log.d("gleeex1",input.length.toString())
//        Log.d("gleeex2",input.peekPosition.toString())
//        Log.d("gleeex3",input.peekPosition.toString())
//        if (header == null) {
//            header = readHeader(input)
//        }
//        track.format(Format.createAudioSampleFormat(
//                null,
//                MimeTypes.AUDIO_RAW,
//                null,
//
//                ))
//        track.format()
//        decoder.queueInputBuffer()
        return 0
    }


    override fun seek(position: Long, timeUs: Long) {

    }

    override fun release() {

    }
}